pallets/multi-asset-delegation/fuzzer/call.rs

```diff
@@ -0,0 +1,364 @@
+		},
+		"whitelist_blueprint_for_rewards" => {
+			// Whitelist blueprint for rewards
+			let is_root = rng.gen_bool(0.5);
+			let (origin, who) = if is_root {
+				(RuntimeOrigin::root(), [0u8; 32].into())
+			} else {
+				random_signed_origin(&mut rng)
+			};
+			fund_account(&mut rng, &who);
+			let blueprint_id = rng.gen::<u64>();
+			(mad::Call::whitelist_blueprint_for_rewards { blueprint_id }, origin)
+		},
+		"manage_asset_in_vault" => {
+			// Manage asset in vault
+			let is_root = rng.gen_bool(0.5);
+			let (origin, who) = if is_root {
+				(RuntimeOrigin::root(), [0u8; 32].into())
+			} else {
+				random_signed_origin(&mut rng)
+			};
+			fund_account(&mut rng, &who);
+			let asset_id = random_asset(&mut rng);
+			let vault_id = rng.gen();
+			let action = if rng.gen() { AssetAction::Add } else { AssetAction::Remove };
+			(mad::Call::manage_asset_in_vault { asset_id, vault_id, action }, origin)
+		},
+		"add_blueprint_id" => {
+			// Add blueprint ID
+			let is_root = rng.gen_bool(0.5);
+			let (origin, who) = if is_root {
+				(RuntimeOrigin::root(), [0u8; 32].into())
+			} else {
+				random_signed_origin(&mut rng)
+			};
+			fund_account(&mut rng, &who);
+			let blueprint_id = rng.gen::<u64>();
+			(mad::Call::add_blueprint_id { blueprint_id }, origin)
+		},
+		"remove_blueprint_id" => {
+			// Remove blueprint ID
+			let is_root = rng.gen_bool(0.5);
+			let (origin, who) = if is_root {
+				(RuntimeOrigin::root(), [0u8; 32].into())
+			} else {
+				random_signed_origin(&mut rng)
+			};
+			fund_account(&mut rng, &who);
+			let blueprint_id = rng.gen::<u64>();
+			(mad::Call::remove_blueprint_id { blueprint_id }, origin)
+		},
+		_ => {
+			unimplemented!("unknown call name: {}", op)
+		},
+	}
+}
+
+fn main() {
+	sp_tracing::try_init_simple();
+	let mut ext = sp_io::TestExternalities::new_empty();
+	let mut events_histogram = Vec::<(mad::Event<Runtime>, u32)>::default();
+	let mut iteration = 0 as BlockNumber;
+	let mut ok = 0;
+	let mut err = 0;
+
+	ext.execute_with(|| {
+		System::set_block_number(1);
+		Session::on_initialize(1);
+		<Staking as Hooks<u64>>::on_initialize(1);
+	});
+
+	loop {
+		fuzz!(|seed: [u8; 32]| {
+			use ::rand::{rngs::SmallRng, SeedableRng};
+			let mut rng = SmallRng::from_seed(seed);
+
+			ext.execute_with(|| {
+				let (call, origin) = random_call(&mut rng);
+				let outcome = call.clone().dispatch_bypass_filter(origin.clone());
+				iteration += 1;
+				match outcome {
+					Ok(_) => ok += 1,
+					Err(_) => err += 1,
+				};
+
+				sp_tracing::trace!(
+					%iteration,
+					?call,
+					?origin,
+					?outcome,
+					%ok,
+					%err,
+					"fuzzed call"
+				);
+
+				// execute sanity checks at a fixed interval, possibly on every block.
+				if iteration
+					% (std::env::var("SANITY_CHECK_INTERVAL")
+						.ok()
+						.and_then(|x| x.parse::<u64>().ok()))
+					.unwrap_or(1) == 0
+				{
+					sp_tracing::info!("running sanity checks at {}", iteration);
+					// TODO
```

Comment:
this is great, but I think we need more tests to ensure that we never end up in a wrong state/ and the math is always correct

Reply:
That is what I thought we should be doing here, however I'm not sure how we should approach that? Any idea would be helpful

Reply:
Checking if the call was ok, and then match over the call and check the state Change? 

Wdyt?

Reply:
Added Sanity checks and improved the overall function calls to be more realistic.